text/0002-observables.md

```diff
@@ -0,0 +1,461 @@
+    const values = [];
+
+    const unsubscribers = stores.map((store, i) => store.subscribe(value => {
+      values[i] = value;
+      if (inited) set(fn(...values));
+    }));
+
+    inited = true;
+    set(fn(...values));
+
+    return function stop() {
+      unsubscribers.forEach(fn => fn());
+    };
+  });
+}
+```
+
+> In the example above, `total` is recalculated immediately whenever the values of `a`, `b` or `c` are set. In some situations that's undesirable; you want to be able to set `a`, `b` *and* `c` without `total` being recalculated until you've finished. That could be done by putting the `set(fn(...values))` in a microtask, but that has drawbacks too. (Of course, that could be a decision left to the user.) Is this a fatal flaw in the design — should we strive for pull-based rather than push-based derived values? Or is it fine in reality?
+
+Derived stores are, by nature, also read-only. They could be used, for example, to filter the items in a todo list:
+
+```html
+<script>
+  import { writable, derive } from 'svelte/store.js';
+  import { todos } from './stores.js';
+
+  const hideDone = writable(false);
+
+  const filtered = derive([todos, hideDone], (todos, hideDone) => todos.filter(todo => {
+    return hideDone ? !todo.done : true;
+  }));
+</script>
+
+<label>
+  <input type=checkbox checked={$hideDone} on:change="{e => hideDone.set(e.target.checked)}">
+  hide done
+</label>
+
+{#each $filtered as todo}
+  <p class="{todo.done ? 'faded' : ''}">{todo.description}</p>
+{/each}
+```
+
+
+### Relationship with TC39 Observables
+
+There is a [stage 1 proposal for an Observable object](https://github.com/tc39/proposal-observable) in JavaScript itself.
+
+Cards on the table: I'm not personally a fan of Observables. I've found them to be confusing and awkward to work with. But there are particular reasons why I don't think they're a good general solution for representing reactive values in a component:
+
+* They don't represent a single value changing over time, but rather a stream of distinct values. This is a subtle but important distinction
+* Two different subscribers to the same Observable could receive different values (!), where as in a UI you want two references to the same value to be guaranteed to be consistent
+* Observables can 'complete', but declarative components (in Svelte and other frameworks) deliberately do not have a concept of time. The two things are incompatible
+* They have error-handling semantics that are very often redundant (what error could occur when observing the mouse position, for example?). When they're not redundant (e.g. in the case of data coming over the network), errors are perhaps best handled out-of-band, since the goal is to concisely represent the value in a component template
+
+Of course, some Observables *are* suitable for representing reactive values in a template, and they could easily be adapted to work with this design:
+
+```js
+function adaptor(observable) {
```

Comment:
This is not really germane to reactive stores in particular, but I've been thinking about how to best handle async values in an inherently sync reactive store (the `subscribe` callback is called immediately with the current value). I think the following may match TC39 observables a little closer (initially async value and handles errors):

```js
const pending = () => new Promise(() => {});
const ok = value => Promise.resolve(value);
const err = error => Promise.reject(error);

function adaptor(observable) {
  return readable(set => {
    const subscription = observable.subscribe({
      next: value => set(ok(value)),
      error: error => set(err(error))
    });

    return subscription.unsubscribe;
  }, pending());
}
```

Then you use `{#await $value}...` to work with them in templates. The subsequent values and errors could be set directly (rather than wrapped in `Promise.resolve`/`reject`), but I think it's better to stay consistent and use promises throughout and it allows for error checking.